### PR TITLE
feat(bitswap): refactor & improve general strategy

### PR DESCRIPTION
**Pull request by dignifiedquire**
Strategy:
- always require to provide a list of `providers`
	- only send wantlists to these providers
- keep a list of `Query`s that deal with the state of work
- `Session`s track the state of individual peers
- drop support for `bitswap 1.0.0`

### CODE REVIEW

File: iroh-bitswap/src/behaviour.rs [hunk 1]
```diff
@@ -202,15 +198,14 @@ impl NetworkBehaviour for Bitswap {
         _failed_addresses: Option<&Vec<Multiaddr>>,
         other_established: usize,
     ) {
-        debug!("connected {} ({})", peer_id, other_established);
-        if other_established > 0 {
-            return;
+        // debug!("connected to {}", peer_id);
```
Comment (ramfox) [inline iroh-bitswap/src/behaviour.rs:201]: stray comment

File: iroh-bitswap/src/behaviour.rs [hunk 2]
```diff
@@ -222,45 +217,83 @@ impl NetworkBehaviour for Bitswap {
         _handler: <Self::ConnectionHandler as IntoConnectionHandler>::Handler,
         remaining_established: usize,
     ) {
-        debug!("disconnected {} ({})", peer_id, remaining_established);
+        // debug!("conn closed: {}, {:?}", peer_id, self.sessions.get(peer_id));
```
Comment (ramfox) [inline iroh-bitswap/src/behaviour.rs:220]: stray comment

File: iroh-bitswap/src/message.rs [hunk 3]
```diff
@@ -16,31 +16,144 @@ mod pb {
...
-/// A bitswap message.
-#[derive(Default, Clone, PartialEq)]
-pub struct BitswapMessage {
+#[derive(Default, Clone, Debug, PartialEq)]
+pub struct Wantlist {
     /// Wanted blocks.
-    pub(crate) want: HashMap<Cid, Priority>,
+    want_blocks: AHashMap<Cid, Priority>,
+    /// Wanted haves.
+    want_haves: AHashSet<Cid>,
```
Comment (ramfox) [inline iroh-bitswap/src/message.rs:24]: tbh this is a pretty confusing (but logical) name. Maybe the comment should read something more like `/// Set of Cids for blocks that are on the want list, but that we already have`?

**Review by ramfox — APPROVED** [review]
looks like the `iroh-bitswap/src/queries.rs` file is empty, probs should be removed, and a couple of instances of dead code

Looks good to me. An interesting way to deal with multiple providers that can have multiple sessions for each provider and multiple queries!

File: iroh-bitswap/src/message.rs (hunk 3 above)
Comment (dignifiedquire) [reply iroh-bitswap/src/message.rs]: removed it for now as we are not using it anyway and it is a very confusing thing

**Pull request merged**